inputremapper/injection/macros/macro.py

```diff
@@ -209,40 +213,7 @@ def __init__(self, code, context):
         self.child_macros = []
 
         self.keystroke_sleep_ms = None
-
-        self._new_event_arrived = asyncio.Event()
-        self._newest_event = None
-        self._newest_action = None
-
-    def notify(self, event, action):
-        """Tell the macro about the newest event."""
-        for macro in self.child_macros:
-            macro.notify(event, action)
-
-        self._newest_event = event
-        self._newest_action = action
-        self._new_event_arrived.set()
-
-    async def _wait_for_event(self, filter=None):
-        """Wait until a specific event arrives.
-
-        The parameters can be used to provide a filter. It will block
-        until an event arrives that matches them.
-
-        Parameters
-        ----------
-        filter : function
-            Receives the event. Stop waiting if it returns true.
-        """
-        while True:
-            await self._new_event_arrived.wait()
-            self._new_event_arrived.clear()
-
-            if filter is not None:
-                if not filter(self._newest_event, self._newest_action):
-                    continue
-
-            break
```

Comment:
There is a whole lot of stuff gone from the macros now. I didn't feel like notifying each macro about every event and then also waiting for a few asyncio Iterations. 

so I rewrote the `add_if_single()` function. Now the context keeps track of the `last_btn_down` and `last_btn_up` event.
`if_single` now waits until the release event or the timeout and compares the `triggering_event` with the `last_btn_down` from the context.

Can you please test with a few macros, especially the `if_single` function? It should be all the same except the `if_single` function now injects after the release or timeout.

Reply:
I'll do later

Reply:
> except the if_single function now injects after the release

yeah, I just wanted to point that out. Other than that it seems to work fine

Reply:
I am not sure if can do anything about that. Do you think this is a huge issue? 
Considering that it only effects the execution of the `other` macro and only if it is not executed by the timeout.

Reply:
Some people expect this behavior (https://github.com/sezanzeb/input-remapper/issues/183). `if_single` was specifically designed for this use-case.

Maybe you can move `_wait_for_event` to the context, so that `if_single` can use it like it used to

Reply:
There might also be a use case for a macro that waits for a specific key in the future. Like `wait_for(KEY_A)` or something. I think it would be good if it would be theoretically possible to more or less easily add something like this.

If macros can listen for events (like `.connect` in gtk, `addEventListener` in javascript or ) it might result in a design that works well for this.

That reminds me, I think connecting to events is better design than passing callback functions. Maybe something like https://www.joeltok.com/blog/2021-3/building-an-event-bus-in-python

Reply:
I reworked the if_single function again. Please have a look at it. It should now work as intended.

Reply:
well ...
Except for some race conditions. I Guess this is why there is a small delay in the keycode_mapper

Reply:
```c#
if_single(
  key(a),
  key(b)
)
```

never releases the b. It says "Injector coroutines ended"

```
16:43:14.903511 3299005 Service INFO config.py:247: Loaded config from "/home/mango/.config/input-remapper/config.json"
16:43:14.904232 3299005 Service DEBUG daemon.py:267: Refreshing because last info is too old
16:43:15.005099 3299005 Service DEBUG groups.py:336: Discovering device paths
16:43:15.005685 3299005 Service DEBUG groups.py:377: Found "3_1_1_input-remapper", "/dev/input/event15", "input-remapper mouse", type: mouse
16:43:15.006021 3299005 Service DEBUG groups.py:377: Found "3_1_1_input-remapper", "/dev/input/event14", "input-remapper gamepad", type: gamepad
16:43:15.006624 3299005 Service DEBUG groups.py:377: Found "3_1_1_input-remapper", "/dev/input/event13", "input-remapper keyboard", type: keyboard
16:43:15.006915 3299005 Service DEBUG groups.py:377: Found "3_7119_5_usb-0000:28:00.3-4.3", "/dev/input/event5", "USB Optical Mouse", type: mouse
16:43:15.007219 3299005 Service DEBUG groups.py:377: Found "3_1118_2048_usb-0000:28:00.3-4.1", "/dev/input/event12", "Microsoft Microsoft® Nano Transceiver v2.0 System Control", type: unknown
16:43:15.007691 3299005 Service DEBUG groups.py:377: Found "3_1118_2048_usb-0000:28:00.3-4.1", "/dev/input/event11", "Microsoft Microsoft® Nano Transceiver v2.0 Consumer Control", type: keyboard
16:43:15.008119 3299005 Service DEBUG groups.py:377: Found "3_1118_2048_usb-0000:28:00.3-4.1", "/dev/input/event9", "Microsoft Microsoft® Nano Transceiver v2.0 Mouse", type: mouse
16:43:15.008493 3299005 Service DEBUG groups.py:377: Found "3_1118_2048_usb-0000:28:00.3-4.1", "/dev/input/event8", "Microsoft Microsoft® Nano Transceiver v2.0", type: keyboard
16:43:15.009048 3299005 Service DEBUG groups.py:377: Found "3_1133_49960_usb-0000:03:00.0-1", "/dev/input/event4", "Logitech USB Keyboard System Control", type: unknown
16:43:15.009433 3299005 Service DEBUG groups.py:377: Found "3_1133_49960_usb-0000:03:00.0-1", "/dev/input/event3", "Logitech USB Keyboard Consumer Control", type: unknown
16:43:15.009835 3299005 Service DEBUG groups.py:377: Found "3_1133_49960_usb-0000:03:00.0-1", "/dev/input/event2", "Logitech USB Keyboard", type: keyboard
16:43:15.010552 3299005 Service INFO groups.py:448: Found "input-remapper mouse", "USB Optical Mouse", "Microsoft Microsoft® Nano Transceiver v2.0", "Logitech USB Keyboard"
16:43:15.010805 3299005 Service INFO mapping.py:179: Loading preset from "/home/mango/.config/input-remapper/presets/Logitech USB Keyboard/new preset 2.json"
16:43:15.011006 3299005 Service DEBUG mapping.py:218: Key((1, 2, 1),) maps to ('if_single(\n  key(a),\n  key(b)\n)', 'keyboard')
16:43:15.014825 3299005 Service INFO injector.py:161: Stopping injecting keycodes for group "Logitech USB Keyboard"
16:43:15.038459 3299005 Service DEBUG daemon.py:461: Using keycodes from "/home/mango/.config/input-remapper/xmodmap.json"
16:43:15.038890 3299005 Service DEBUG system_mapping.py:144: Updated keycodes with 0 new ones
16:43:15.041251 3299704 Service INFO injector.py:300: Starting injecting the mapping for "Logitech USB Keyboard"
16:43:15.042482 3299704 Service DEBUG parse.py:416: preparing macro if_single(key(a),key(b)) for later execution
16:43:15.042841 3299704 Service DEBUG parse.py:252: calls if_single with key(a),key(b)
16:43:15.043173 3299704 Service DEBUG parse.py:252:   calls key with a
16:43:15.043293 3299704 Service DEBUG parse.py:309:     string a
16:43:15.043381 3299704 Service DEBUG parse.py:273:   add call to key with ['a'], {}
16:43:15.043713 3299704 Service DEBUG parse.py:252:   calls key with b
16:43:15.043823 3299704 Service DEBUG parse.py:309:     string b
16:43:15.043910 3299704 Service DEBUG parse.py:273:   add call to key with ['b'], {}
16:43:15.044044 3299704 Service DEBUG parse.py:273: add call to if_single with [<Macro "key(a)">, <Macro "key(b)">], {}
16:43:15.044249 3299704 Service DEBUG mapping_parser.py:76: created mapping handler:
16:43:15.044401 3299704 Service DEBUG mapping_parser.py:77: CombinationHandler for ((1, 2, 1),) <140576378363376>:
16:43:15.044489 3299704 Service DEBUG mapping_parser.py:77:     MacroHandler <140576378361552>:
16:43:15.044585 3299704 Service DEBUG mapping_parser.py:77:         maps to <Macro "if_single(key(a),key(b))"> on keyboard
16:43:15.044819 3299704 Service DEBUG injector.py:210: No need to grab /dev/input/event4
16:43:15.045255 3299704 Service DEBUG injector.py:210: No need to grab /dev/input/event3
16:43:15.045570 3299704 Service DEBUG injector.py:197: Grabbing "/dev/input/event2" because of "Key((1, 2, 1),)"
16:43:15.045749 3299704 Service DEBUG injector.py:217: Grab /dev/input/event2
16:43:15.179320 3299704 Service DEBUG consumer_control.py:94: Starting to listen for events from /dev/input/event2, fd 44
16:43:16.886627 3299704 Service DEBUG mapping_handler.py:160: Key((1, 2, 1)) ---------------- triggered: sending to sub-handler
16:43:17.190412 3299704 Service DEBUG mapping_handler.py:277: (1, 48, 1) -------------------- sending from macro to keyboard
16:43:17.190733 3299704 Service DEBUG injector.py:361: Injector coroutines ended
```

Reply:
that is new, now I have to go back and see when I broke it

Reply:
should be fixed now